### PR TITLE
Implement ArrayInput field

### PR DESCRIPTION
**Pull request by mudrila**
- Create `ArrayInput` component
- Adjust logic of validation and argument preparation
- Adjust rendering logic to display `ArrayInput` component in proper cases

### CODE REVIEW

File: src/components/Controls/ArrayInput/hooks/index.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	const handleInputChange: InputChangeListener = event => {
+		setInputValue(event.target.value)
+	}
+
+	const handleKeyDown: KeyboardEventHandler<HTMLDivElement> = event => {
+		if (!inputValue) return
+		switch (event.key) {
+			case "Enter":
+			case "Tab":
+				const newValue = [...value, inputValue]
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/hooks/index.ts:30]: This is not consistent. You have onRemove(index) handler below, so, first, it's more obvious to call this handler onAdd instead of onChange, second, just call it with the added element instead of constructing the whole array here.
Imagine, we could do something with added element before updating the array in the parent component, and this way doesn't allow to see which element is added.

File: src/components/Controls/ArrayInput/hooks/index.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+				setInputValue("")
+				onChange(newValue)
+				event.preventDefault()
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/hooks/index.ts:33]: This is more logical to be called first

File: src/components/Controls/ArrayInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	placeholder?: string
+}> = ({
+	onChange,
+	onRemove,
+	value,
+	borders,
+	validation,
+	placeholder = "Type value and press enter or tab..."
+}) => {
+	const {handleInputChange, handleKeyDown, inputValue} = useArrayInput({
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:26]: And I don't think we ever need this hook. It's not reusable and used only to organize the logic (which is good, in general), however, this is too small component and there's of no use to move out 2 functions.

File: src/components/Controls/ArrayInput/index.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		onChange,
+		value
+	})
+	return (
+		<section className="array-input-container">
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:31]: First, please use BEM naming methodology for classes.
Second, why it is a section? Section is intended to be a large section of the page.

File: src/components/Controls/ArrayInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import {FunctionComponent} from "react"
+import CloseIcon from "../../../assets/icons/CloseIcon"
+import Input from "../Input"
+import Button from "../Button"
+import useArrayInput from "./hooks"
+import "./styles.scss"
+
+export type ArrayInputChangeListener = (newValue: ArrayInputValue) => void
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:8]: Again, why do you need to define these types?

File: src/components/Controls/ArrayInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		value
+	})
+	return (
+		<section className="array-input-container">
+			<Input
+				onChange={handleInputChange}
+				value={inputValue}
+				onKeyDown={handleKeyDown}
+				placeholder={placeholder}
+				borders={borders || "all"}
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:37]: I'd better use default value in the component args. You are already doing it for placeholder, why not to keep it consistent?
Other option is to make this arg mandatory

File: src/components/Controls/ArrayInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+				validation={validation}
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:38]: If you are using validation, add the logic which will prevent new element from adding if validation is not nullish

File: src/components/Controls/ArrayInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+			/>
+			{value.map((option, index) => (
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:40]: IMO value is not the best name as it sounds like a scalar. You can call it elements or tags, so the name stands in plural form.

File: src/components/Controls/ArrayInput/index.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+				<section key={option} className="array-input-container__option">
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:41]: This will cause a warning if we have same strings as arguments. Use index instead.

File: src/components/Controls/ArrayInput/index.tsx (hunk 4 above)
Comment (Koochr) [reply src/components/Controls/ArrayInput/index.tsx:31]: Upd, I see you are using BEM, but why such a long name? Why not just array-input?

File: src/components/Controls/ArrayInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+					<span>{option}</span>
+					<Button
+						buttonType="secondary"
+						onClick={() => onRemove(index)}
+						extraClassName="array-input-container__option__remove-button"
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/index.tsx:46]: This name also doesn't match BEM.
should be block__element--modifier (the last is optional)

File: src/components/Controls/ArrayInput/styles.scss [hunk 11]
```diff
@@ -0,0 +1,32 @@
+@import "src/assets/styles/variables";
+
+.array-input-container {
+    font-size: 16px;
+    line-height: 20px;
+    input {
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/styles.scss:6]: Did you use stylelint? Should be empty lines here.
If it doesn't fix it, please do some research if it can.

File: src/components/Controls/ArrayInput/styles.scss [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    &__option {
+        padding: 5px;
+        margin-top: 10px;
+        margin-right: 5px;
+        border: 1px solid $black;
+        display: inline-flex;
+        align-items: center;
+        justify-content: center;
+        & &__remove-button {
```
Comment (Koochr) [inline src/components/Controls/ArrayInput/styles.scss:17]: Why 2 & symbols?

File: src/components/Controls/ArrayInput/index.tsx (hunk 5 above)
Comment (Koochr) [reply src/components/Controls/ArrayInput/index.tsx:8]: We keep all the types in `types` folder. However, these ones are too simple to define.

File: src/utlls/index.ts [hunk 13]
```diff
@@ -31,8 +31,8 @@ export const formatTimeDifference = (ms: number): string => {
 	return `${hours > 0 ? `${hours} h ` : ""}${minutes} m`
 }
 
-const validateScalarArgument = (arg: string, argType: AbiDataType): boolean => {
-	if (argType.endsWith("[]")) {
+const validateScalarArgument = (arg: string | readonly string[], argType: AbiDataType): boolean => {
```
Comment (Koochr) [inline src/utlls/index.ts:34]: This function is not intended to accept an array. Why did you change the input type?

File: src/utlls/index.ts [hunk 14]
```diff
@@ -89,16 +91,14 @@ const prepareScalarArgument = (arg: string, dataType: AbiScalar): string | boole
...
 	dataTypes: AbiDataType[]
 ): (string | string[] | boolean | boolean[] | BigNumber | BigNumber[])[] =>
 	args.map((arg, index) => {
-		if (dataTypes[index].endsWith("[]")) {
-			return arg
-				.split(",")
-				.map(splittedArg =>
-					prepareScalarArgument(splittedArg, dataTypes[index].slice(0, -2) as AbiScalar)
-				) as string[] | boolean[] | BigNumber[]
+		if ((dataTypes[index].endsWith("[]") && arg instanceof Array) || arg instanceof Array) {
```
Comment (Koochr) [inline src/utlls/index.ts:98]: This can be simplified to `dataTypes[index].endsWith("[]") && arg instanceof Array`

**Review by Koochr — COMMENTED** [review]
In general, looks good, however, there are various minor things to fix/improve.
I might be nagging regarding some cosmetic things like type declarations, however, I mention this to keep some consistency across the project code style.

File: src/utlls/index.ts (hunk 14 above)
Comment (mudrila) [reply src/utlls/index.ts:98]: That's TS compatibility - it will complain about final return, cause if `arg instanceof Array` but `(dataTypes[index].endsWith("[]")` is false (which can't be true, we know - but TS don't know that) then TS assumes that array might drop into `prepareScalarArgument`. The best option could be to use just `arg instanceof Array`, I believe
